Analysis/LeptonID/src/PIDutil.cc

```diff
@@ -0,0 +1,374 @@
+    dist_f.SetParameter(14, momentumOfLine[1]);
+    dist_f.SetParameter(15, momentumOfLine[2]);
+
+    // TODO: Yasser had an extra SetRange call, is this needed?
+
+    double phi = phi_zero;
+    double v = 0.0;
+
+    double distance = dist_f.GetMinimumXY(phi, v);
+
+    double xTrackPCA = x_ref[0] + (invOmega - track->getD0()) * sin(phi_zero) - sin(phi) * invOmega;
+    double yTrackPCA = x_ref[1] - (invOmega - track->getD0()) * cos(phi_zero) + cos(phi) * invOmega;
+    double zTrackPCA = x_ref[2] + track->getZ0() - (phi-phi_zero) * track->getTanLambda() * invOmega;
+
+    PCAatTrack = TVector3(xTrackPCA, yTrackPCA, zTrackPCA);
+
+    double xLine = pointOnLine[0] + momentumOfLine.Px() * v;
+	double yLine = pointOnLine[1] + momentumOfLine.Py() * v;
+	double zLine = pointOnLine[2] + momentumOfLine.Pz() * v;
+
+    PCAatLine = TVector3(xLine, yLine, zLine);
+
+    return distance;
+}
+
+const MCParticle *checkSLD(const MCParticle *mcp, const MCParticle *parent)
+{
+    if (mcp == nullptr || parent == nullptr)
+    {
+        return nullptr;
+    }
+
+    int parentPDG = abs(parent->getPDG());
+    bool isCHadron = parentPDG / 100 == 4 || parentPDG / 1000 == 4;
+    bool isBHadron = parentPDG / 100 == 5 || parentPDG / 1000 == 5;
+
+    if (!(isCHadron || isBHadron))
+    {
+        return nullptr;
+    }
+
+    int mcPDG = mcp->getPDG();
+    int mcPDGsign = (0 < mcPDG) - (mcPDG < 0);
+    // nu pdg is one number higher and it should have opposite sign
+    int nuPDG = -1 * mcPDGsign * (abs(mcPDG) + 1);
+
+    auto &daughters = parent->getDaughters();
+    for (const auto *d : daughters)
+    {
+        if (d->getGeneratorStatus() == 1 && d->getPDG() == nuPDG)
+        {
+            return d;
+        }
+    }
+
+    return nullptr;
+}
+
+VertexVec getJetVertices(const ReconstructedParticle *jet)
+{
+    std::set<Vertex *> jetVertices{};
+    // iterate over jet pfos & save all non prim vertices
+    for (const auto *reco : jet->getParticles())
+    {
+        auto vtx = reco->getStartVertex();
+        if (vtx != nullptr && !vtx->isPrimary())
+        {
+            jetVertices.insert(vtx);
+        }
+    }
+    return VertexVec(jetVertices.begin(), jetVertices.end());
+}
+
+Vertex *getClosestVtx(const Vertex *primVtx, VertexVec jetVertices)
+{
+    Vertex *closestVtx = nullptr;
+
+    const float* primVtxPos = primVtx->getPosition();
+
+    double minDist = 1000000.0;
+    for (auto *vtx : jetVertices)
+    {
+        const float* vtxPos = vtx->getPosition();
+
+        double dist_x_sq = (vtxPos[0] - primVtxPos[0]) * (vtxPos[0] - primVtxPos[0]);
+        double dist_y_sq = (vtxPos[1] - primVtxPos[1]) * (vtxPos[1] - primVtxPos[1]);
+        double dist_z_sq = (vtxPos[2] - primVtxPos[2]) * (vtxPos[2] - primVtxPos[2]);
+
+        double dist = sqrt(dist_x_sq + dist_y_sq + dist_z_sq);
+
+        if (dist < minDist)
+        {
+            minDist = dist;
+            closestVtx = vtx;
+        }
+    }
```

Comment:
This is another candidate for `std::min_element` and a lambda which would make the intent immediately clear. That would also eliminate the need for an explicit `minDist` variable. Additionally the `sqrt` is unnecessary here, since we are only interested in how the distances compare not their actual values.